pkg/controller/master/upgradelog/controller.go

```diff
@@ -0,0 +1,615 @@
+	deploymentClient    ctlappsv1.DeploymentClient
+	jobClient           ctlbatchv1.JobClient
+	jobCache            ctlbatchv1.JobCache
+	loggingClient       ctlloggingv1.LoggingClient
+	managedChartClient  ctlmgmtv3.ManagedChartClient
+	managedChartCache   ctlmgmtv3.ManagedChartCache
+	pvcClient           ctlcorev1.PersistentVolumeClaimClient
+	statefulSetClient   ctlappsv1.StatefulSetClient
+	statefulSetCache    ctlappsv1.StatefulSetCache
+	upgradeClient       ctlharvesterv1.UpgradeClient
+	upgradeCache        ctlharvesterv1.UpgradeCache
+	upgradeLogClient    ctlharvesterv1.UpgradeLogClient
+	upgradeLogCache     ctlharvesterv1.UpgradeLogCache
+}
+
+func (h *handler) OnUpgradeLogChange(_ string, upgradeLog *harvesterv1.UpgradeLog) (*harvesterv1.UpgradeLog, error) {
+	if upgradeLog == nil || upgradeLog.DeletionTimestamp != nil {
+		return upgradeLog, nil
+	}
+	logrus.Infof("Processing UpgradeLog %s/%s", upgradeLog.Namespace, upgradeLog.Name)
+
+	if harvesterv1.UpgradeLogReady.GetStatus(upgradeLog) == "" {
+		logrus.Info("Initialize UpgradeLog")
+		toUpdate := upgradeLog.DeepCopy()
+		harvesterv1.UpgradeLogReady.CreateUnknownIfNotExists(toUpdate)
+		return h.upgradeLogClient.Update(toUpdate)
+	}
+
+	if harvesterv1.OperatorDeployed.GetStatus(upgradeLog) == "" {
+		logrus.Info("Check if there are any existing logging-operator")
+
+		toUpdate := upgradeLog.DeepCopy()
+		harvesterv1.OperatorDeployed.CreateUnknownIfNotExists(toUpdate)
+
+		// Detect rancher-logging Addon
+		if addon, err := h.addonCache.Get(addonNamespace, rancherLoggingAddonName); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return nil, err
+			}
+			logrus.Info("rancher-logging Addon is not installed")
+		} else {
+			if addon.Status.Status == harvesterv1.AddonEnabled {
+				setOperatorDeployedCondition(toUpdate, corev1.ConditionTrue, "Skipped", "rancher-logging Addon is enabled")
+				return h.upgradeLogClient.Update(toUpdate)
+			}
+			logrus.Info("rancher-logging Addon is not enabled")
+		}
+
+		// Detect the rancher-logging ManagedChart
+		if managedChart, err := h.managedChartCache.Get(managedChartNamespace, rancherLoggingManagedChartName); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return nil, err
+			}
+			logrus.Info("rancher-logging ManagedChart is not installed")
+		} else {
+			if managedChart.Status.Summary.DesiredReady > 0 && managedChart.Status.Summary.DesiredReady == managedChart.Status.Summary.Ready {
+				setOperatorDeployedCondition(toUpdate, corev1.ConditionTrue, "Skipped", "rancher-logging ManagedChart is ready")
+				return h.upgradeLogClient.Update(toUpdate)
+			}
+			logrus.Warn("rancher-logging ManagedChart is not ready")
+			return nil, err
+		}
+
+		// If none of the above exists, install the customized rancher-logging ManagedChart
+		logrus.Info("Deploy logging-operator")
+		if _, err := h.managedChartClient.Create(prepareOperator(upgradeLog)); err != nil && !apierrors.IsAlreadyExists(err) {
+			return nil, err
+		}
```

Comment:
Thanks for bringing up a different way of thinking. The causality of "enabling rancher-logging addon" and "enabling logging for upgrades" is a bit implicit since there's no strong binding between them. And we have a `LogEnabled` toggle in the upgrade CRD for users to choose from (it will be `true` by default). It's still better to have the deployment of the logging-operator handled by the upgradelog controller. More complied with the style of declaration.

cc @bk201 @guangbochen